server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/ProjectStatusAction.java

```diff
@@ -155,20 +156,31 @@ private ProjectAndSnapshot getSnapshotThenProject(DbSession dbSession, String an
   }
 
   private SnapshotDto getSnapshot(DbSession dbSession, String analysisUuid) {
-    java.util.Optional<SnapshotDto> snapshotDto = dbClient.snapshotDao().selectByUuid(dbSession, analysisUuid);
+    Optional<SnapshotDto> snapshotDto = dbClient.snapshotDao().selectByUuid(dbSession, analysisUuid);
     return checkFoundWithOptional(snapshotDto, "Analysis with id '%s' is not found", analysisUuid);
   }
 
-  private Optional<String> getQualityGateDetailsMeasureData(DbSession dbSession, ComponentDto project) {
-    MeasureQuery measureQuery = MeasureQuery.builder()
-      .setProjectUuids(singletonList(project.projectUuid()))
-      .setMetricKey(CoreMetrics.QUALITY_GATE_DETAILS_KEY)
-      .build();
-    List<MeasureDto> measures = dbClient.measureDao().selectByQuery(dbSession, measureQuery);
+  private Optional<String> loadQualityGateDetails(DbSession dbSession, ProjectAndSnapshot projectAndSnapshot, boolean onAnalysis) {
+    if (onAnalysis) {
+      if (!projectAndSnapshot.snapshotDto.isPresent()) {
+        return Optional.empty();
+      }
+      // get the gate status as it was computed during the specified analysis
+      String analysisUuid = projectAndSnapshot.snapshotDto.get().getUuid();
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Call "Optional#isPresent()" before accessing the value. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3655)